README
!!! DISCLAIMER !!!

THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


!!! /DISCLAIMER !!!





hashkill-0.2.0
==============

1. What is hashkill?
2. Installation
3. Features
4. Plugins
5. TODO
6. Credits




1. What is hashkill?
====================

hashkill (stupid name I know) is an open-source password cracker that uses the OpenSSL library to 
crack different types of password hashes.

Basically, its purposes are:

* allow cracking lots of different types of password hashes on linux
* employ easy to use command-line interface
* multithreaded
* SSE2/XOP accelerated
* GPU accelerated
* open-sourced, GPL-licensed

Disadvantages at that moment are that it does not support GPU-based cracking. 



2. Installation
===============

Everything boils down to installing the package prerequisites, ./configure && make && make install.

Useful options to configure are:

--datadir : the directory where all plugins, kernels, markov models and preinstalled rulesets will be placed
--bindir  : location of the hashkill binary

Build prerequisites:

* gcc >=4.6
* libc headers
* json-c (also known as libjson0) and its headers for session save/restore support
* proprietary AMD/NVidia drivers for GPU support and kernel compilation
* zlib and its headers
* openssl and its headers
* pkgconfig
* libtool
* make



3. Features
===========

Currently (0.2.4) hashkill supports 4 attack modes:

* Markov attack (default)
* bruteforce
* rule

hashkill supports sessions (e.g interrupted session can be restored)

Currently there are 41 plugins that can crack more than 45 different hash types.

And lots of bugs I guess :)


4. Plugins
==========

The following plugins are currently available:


apr1            Apache apr1 plugin
desunix         DES(Unix) plugin (.htpasswd)
dmg             FileVault (v1)  passwords plugin
ipb2            md5(md5(salt).md5(pass)) plugin (IPB > 2.x)
ldap-sha        LDAP SHA plugin
ldap-ssha       LDAP SSHA (salted SHA) plugin
lm              LM plugin
md4             MD4 plugin
md5-passsalt    md5(password,salt) plugin (joomla)
md5-saltpass    md5(salt,password) plugin (osCommerce)
md5             MD5 plugin
md5md5          md5(md5(pass)) plugin
md5unix         MD5(Unix) plugin (shadow files)
mediawiki       md5(salt.'-'.md5(password)) plugin (Wikimedia)
mscash          Domain cached credentials plugin
mscash2         Domain cached credentials v2 plugin
mssql-2000      Microsoft SQL Server 2000 plugin
mssql-2005      Microsoft SQL Server 2005 plugin
mysql5          MySQL > 4.1 plugin
ntlm            NTLM plugin
oracle-old      Oracle 7 up to 10r2 plugin
oracle11g       Oracle 11g plugin
osx-old         MacOSX <= 10.6 system passwords plugin
osxlion         MacOSX Lion system passwords plugin
phpbb3          phpBB3 hashes plugin
pixmd5          Cisco PIX password hashes plugin
privkey         SSH/SSL private key passphrase plugin
rar             RAR3 passwords plugin
ripemd160       RIPEMD-160 plugin
sapg            SAP CODVN G passwords plugin
sha1            SHA1 plugin
sha1sha1        sha1(sha1(pass)) plugin
sha256          SHA-256 plugin
sha512          SHA-512 plugin
sha512unix      SHA512(Unix) plugin (shadow files)
sl3             Nokia SL3 plugin
smf             SMF plugin
vbulletin       md5(md5(pass).salt) plugin
wordpress       Wordpress hashes plugin
wpa             WPA-PSK plugin
zip             ZIP passwords plugin


For detailed info on a plugin, type:

hashkill -P<plugin> (e.g hashkill -Pmd5)




5. TODO
=======

More plugins


6. Credits
==========

See the CREDITS file


Changelog
2010-04-18  Milen Rangelov <gat3way@gat3way.eu>

    * Created project skeleton - hashinterface, main, threads, plugins, plugin/

    * threads.c Thread synchronization uses semaphores/mutexes instead of condvars/mutexes

    * plugins/md5.c - a simple md5 plugin

    * threads.c (monitor_thread) - shows 4-core per-thread queue/cracked stats


2010-04-23 Milen Rangelov <gat3way@gat3way.eu>

    * hashinterface.c (hashk_md5_hex/hashk_sha1_hex) - implemented optimized rudimentary sprintf(). Performance up 25%

    * threads.c (worker_thread_linear/worker_thread_nonlinear) - split worker thread logic depending on attack complexity 
    If attack has "linear" complexity (no salts) - then we do not remove cracked hashes from hash_list. This avoids mutex
    locking at speeds up cracking considerably.


2010-04-24 Milen Rangelov <gat3way@gat3way.eu>

    * plugins/md5md5.c - md5(md5(password)) plugin

    * plugins/md5-saltpass.c - md5(salt,password) plugin

    * plugins/sha1.c - sha1(password) plugin


2010-04-25 Milen Rangelov <gat3way@gat3way.eu>

    * plugins/mysql5.c - mysql >= 4.1 password hashes plugin
    
    * threads.c (main_thread_bruteforce) - customizable bruteforce


2010-04-26 Milen Rangelov <gat3way@gat3way.eu>

    * autoconf/automake skeleton. 


2010-04-27 Milen Rangelov <gat3way@gat3way.eu>

    * main.c - fixed the option parsing mess

    * released version 0.1.0


2010-05-01 Milen Rangelov <gat3way@gat3way.eu>

    * threads.c (spawn_threads) - fixed race condition with pthread_create(). Released 0.1.1 to address the problem.

    * threads.c (monitor_thread) - display human-readible stats

    * sessions.c/threads.c - monitor_thread() now saves session automatically. 


2010-05-02 Milen Rangelov <gat3way@gat3way.eu>

    * sessions.c/main.c/threads.c - we now save/restore sessions


2010-05-03 Milen Rangelov <gat3way@gat3way.eu>

    * valgrind tests, fixed invalid memory write bug while parsing dictionary filename from option arguments

    * getpwuid() and pthread_create() seem to leak memory wtf?!?

    * ran .c and .h files through splint (weak). Fixed a number of non-important issues


2010-05-05 Milen Rangelov <gat3way@gat3way.eu>

    * sessions.c (session_init_file)- fixed symlink security issue when writing session files.


2010-05-09 Milen Rangelov <gat3way@gat3way.eu>

    * threads.c/main.c/sessions.c - implemented hybrid attack


2010-05-13 Milen Rangelov <gat3way@gat3way.eu>

    * plugins/hashunix.c
    * plugins/sha512.c
    * plugins/sha256.c
    * plugins/smf.c


2010-05-14 Milen Rangelov <gat3way@gat3way.eu>

    * plugins/desunix.c
    * plugins/md5-passalt.c
    * plugins/vbulletin.c
    * threads.c - added OpenSSL crypto callbacks

2010-05-15 Milen Rangelov <gat3way@gat3way.eu>

    * plugins/wordpress.c
    * plugins/phpbb3.c


2010-05-16 Milen Rangelov <gat3way@gat3way.eu>

    * plugins/ipb2.c
    * plugins/mysql-old.c
    * threads.c - removed semaphores code, using only mutexes for sync



2010-05-16 - 2010-06-05 Milen Rangelov <gat3way@gat3way.eu>

    Overall optimizations, attacks run almost 100% faster now.
    Added -o option to save output to a file
    Fixed <enter> key detailed stats
    Created a manpage
    

2010-06-06 Milen Rangelov <gat3way@gat3way.eu>

    Critical fix to session code. Release 0.2.1a
    

2010-06-06 - 2010-07-11 Milen Rangelov <gat3way@gat3way.eu>

    * Added Markov attack, now default
    * plugins/rsa-privkey.c
    * plugins/zip.c
    * plugins/ripemd160.c
    * plugins/sha1sha1.c
    * Long options recognized
    * Various small bugfixes
    
    Released 0.2.2
    

2010-07-11 - 2010-09-15 Milen Rangelov <gat3way@gat3way.eu>

    * Changed default options (markov/rockyou/10). Bruteforce defaults are 1:8:lalphanum
    * plugins/lm.c
    * plugins/ntlm.c
    * plugins/oracle11g.c
    * plugins/oracle-old.c
    * plugins/mssql-2000.c
    * plugins/mssql-2005.c
    * plugins/ldap-sha.c
    * plugins/ldap-ssha.c
    * plugins/apr1.c
    * plugins/cisco-pix.c
    * plugins/md4.c
    * Added SSE2-accelerated md4,md5 and sha1 functions.
    * Added SSE2-accelerated bitslice DES function
    
    Released 0.2.3



src/base64.c
/* base64.c
 *
 * hashkill - a hash cracking tool
 * Copyright (C) 2010 Milen Rangelov <gat3way@gat3way.eu>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


/*
 * Copyright (c) 1996-1999 by Internet Software Consortium.
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND INTERNET SOFTWARE CONSORTIUM DISCLAIMS
 * ALL WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL INTERNET SOFTWARE
 * CONSORTIUM BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 * DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 * PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
 * ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS
 * SOFTWARE.
 */

/*
 * Portions Copyright (c) 1995 by International Business Machines, Inc.
 *
 * International Business Machines, Inc. (hereinafter called IBM) grants
 * permission under its copyrights to use, copy, modify, and distribute this
 * Software with or without fee, provided that the above copyright notice and
 * all paragraphs of this notice appear in all copies, and that the name of IBM
 * not be used in connection with the marketing of any product incorporating
 * the Software or modifications thereof, without specific, written prior
 * permission.
 *
 * To the extent it has a right to do so, IBM grants an immunity from suit
 * under its patents, if any, for the use, sale or manufacture of products to
 * the extent that such products are used for performing Domain Name System
 * dynamic updates in TCP/IP networks by means of the Software.  No immunity is
 * granted for any product per se or for any other function of any product.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", AND IBM DISCLAIMS ALL WARRANTIES,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE.  IN NO EVENT SHALL IBM BE LIABLE FOR ANY SPECIAL,
 * DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE, EVEN
 * IF IBM IS APPRISED OF THE POSSIBILITY OF SUCH DAMAGES.
 */



#include <sys/types.h>
#define u_char unsigned char

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


#define Assert(Cond) if (!(Cond)) abort()

static const char Base64[] =
	"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
static const char Pad64 = '=';

/* (From RFC1521 and draft-ietf-dnssec-secext-03.txt)
   The following encoding technique is taken from RFC 1521 by Borenstein
   and Freed.  It is reproduced here in a slightly edited form for
   convenience.

   A 65-character subset of US-ASCII is used, enabling 6 bits to be
   represented per printable character. (The extra 65th character, "=",
   is used to signify a special processing function.)

   The encoding process represents 24-bit groups of input bits as output
   strings of 4 encoded characters. Proceeding from left to right, a
   24-bit input group is formed by concatenating 3 8-bit input groups.
   These 24 bits are then treated as 4 concatenated 6-bit groups, each
   of which is translated into a single digit in the base64 alphabet.

   Each 6-bit group is used as an index into an array of 64 printable
   characters. The character referenced by the index is placed in the
   output string.

                         Table 1: The Base64 Alphabet

      Value Encoding  Value Encoding  Value Encoding  Value Encoding
          0 A            17 R            34 i            51 z
          1 B            18 S            35 j            52 0
          2 C            19 T            36 k            53 1
          3 D            20 U            37 l            54 2
          4 E            21 V            38 m            55 3
          5 F            22 W            39 n            56 4
          6 G            23 X            40 o            57 5
          7 H            24 Y            41 p            58 6
          8 I            25 Z            42 q            59 7
          9 J            26 a            43 r            60 8
         10 K            27 b            44 s            61 9
         11 L            28 c            45 t            62 +
         12 M            29 d            46 u            63 /
         13 N            30 e            47 v
         14 O            31 f            48 w         (pad) =
         15 P            32 g            49 x
         16 Q            33 h            50 y

   Special processing is performed if fewer than 24 bits are available
   at the end of the data being encoded.  A full encoding quantum is
   always completed at the end of a quantity.  When fewer than 24 input
   bits are available in an input group, zero bits are added (on the
   right) to form an integral number of 6-bit groups.  Padding at the
   end of the data is performed using the '=' character.

   Since all base64 input is an integral number of octets, only the
         -------------------------------------------------                       
   following cases can arise:
   
       (1) the final quantum of encoding input is an integral
           multiple of 24 bits; here, the final unit of encoded
	   output will be an integral multiple of 4 characters
	   with no "=" padding,
       (2) the final quantum of encoding input is exactly 8 bits;
           here, the final unit of encoded output will be two
	   characters followed by two "=" padding characters, or
       (3) the final quantum of encoding input is exactly 16 bits;
           here, the final unit of encoded output will be three
	   characters followed by one "=" padding character.
   */

int
b64_ntop(u_char const *src, size_t srclength, char *target, size_t targsize) {
	size_t datalength = 0;
	u_char input[3];
	u_char output[4];
	size_t i;

	while (2 < srclength) {
		input[0] = *src++;
		input[1] = *src++;
		input[2] = *src++;
		srclength -= 3;

		output[0] = input[0] >> 2;
		output[1] = ((input[0] & 0x03) << 4) + (input[1] >> 4);
		output[2] = ((input[1] & 0x0f) << 2) + (input[2] >> 6);
		output[3] = input[2] & 0x3f;
		Assert(output[0] < 64);
		Assert(output[1] < 64);
		Assert(output[2] < 64);
		Assert(output[3] < 64);

		if (datalength + 4 > targsize)
			return (-1);
		target[datalength++] = Base64[output[0]];
		target[datalength++] = Base64[output[1]];
		target[datalength++] = Base64[output[2]];
		target[datalength++] = Base64[output[3]];
	}
    
	/* Now we worry about padding. */
	if (0 != srclength) {
		/* Get what's left. */
		input[0] = input[1] = input[2] = '\0';
		for (i = 0; i < srclength; i++)
			input[i] = *src++;
	
		output[0] = input[0] >> 2;
		output[1] = ((input[0] & 0x03) << 4) + (input[1] >> 4);
		output[2] = ((input[1] & 0x0f) << 2) + (input[2] >> 6);
		Assert(output[0] < 64);
		Assert(output[1] < 64);
		Assert(output[2] < 64);

		if (datalength + 4 > targsize)
			return (-1);
		target[datalength++] = Base64[output[0]];
		target[datalength++] = Base64[output[1]];
		if (srclength == 1)
			target[datalength++] = Pad64;
		else
			target[datalength++] = Base64[output[2]];
		target[datalength++] = Pad64;
	}
	if (datalength >= targsize)
		return (-1);
	target[datalength] = '\0';	/* Returned value doesn't count \0. */
	return (datalength);
}

/* skips all whitespace anywhere.
   converts characters, four at a time, starting at (or after)
   src from base - 64 numbers into three 8 bit bytes in the target area.
   it returns the number of data bytes stored at the target, or -1 on error.
 */

int
b64_pton(src, target, targsize)
	char const *src;
	u_char *target;
	size_t targsize;
{
	int tarindex, state, ch;
	char *pos;

	state = 0;
	tarindex = 0;

	while ((ch = *src++) != '\0') {
		if (isspace(ch))	/* Skip whitespace anywhere. */
			continue;

		if (ch == Pad64)
			break;

		pos = strchr(Base64, ch);
		if (pos == 0) 		/* A non-base64 character. */
			return (-1);

		switch (state) {
		case 0:
			if (target) {
				if ((size_t)tarindex >= targsize)
					return (-1);
				target[tarindex] = (pos - Base64) << 2;
			}
			state = 1;
			break;
		case 1:
			if (target) {
				if ((size_t)tarindex + 1 >= targsize)
					return (-1);
				target[tarindex]   |=  (pos - Base64) >> 4;
				target[tarindex+1]  = ((pos - Base64) & 0x0f)
							<< 4 ;
			}
			tarindex++;
			state = 2;
			break;
		case 2:
			if (target) {
				if ((size_t)tarindex + 1 >= targsize)
					return (-1);
				target[tarindex]   |=  (pos - Base64) >> 2;
				target[tarindex+1]  = ((pos - Base64) & 0x03)
							<< 6;
			}
			tarindex++;
			state = 3;
			break;
		case 3:
			if (target) {
				if ((size_t)tarindex >= targsize)
					return (-1);
				target[tarindex] |= (pos - Base64);
			}
			tarindex++;
			state = 0;
			break;
		default:
			abort();
		}
	}

	/*
	 * We are done decoding Base-64 chars.  Let's see if we ended
	 * on a byte boundary, and/or with erroneous trailing characters.
	 */

	if (ch == Pad64) {		/* We got a pad char. */
		ch = *src++;		/* Skip it, get next. */
		switch (state) {
		case 0:		/* Invalid = in first position */
		case 1:		/* Invalid = in second position */
			return (-1);

		case 2:		/* Valid, means one byte of info */
			/* Skip any number of spaces. */
			for ((void)NULL; ch != '\0'; ch = *src++)
				if (!isspace(ch))
					break;
			/* Make sure there is another trailing = sign. */
			if (ch != Pad64)
				return (-1);
			ch = *src++;		/* Skip the = */
			/* Fall through to "single trailing =" case. */
			/* FALLTHROUGH */

		case 3:		/* Valid, means two bytes of info */
			/*
			 * We know this char is an =.  Is there anything but
			 * whitespace after it?
			 */
			for ((void)NULL; ch != '\0'; ch = *src++)
				if (!isspace(ch))
					return (-1);

			/*
			 * Now make sure for cases 2 and 3 that the "extra"
			 * bits that slopped past the last full byte were
			 * zeros.  If we don't check them, they become a
			 * subliminal channel.
			 */
			if (target && target[tarindex] != 0)
				return (-1);
		}
	} else {
		/*
		 * We ended by seeing the end of the string.  Make sure we
		 * have no partial bytes lying around.
		 */
		if (state != 0)
			return (-1);
	}

	return (tarindex);
}

